plugins/zmq/src/plugin.cpp

```diff
@@ -0,0 +1,447 @@
+                                       error.num()));
+  }
+
+  static auto poll(::zmq::socket_t& socket, short flags,
+                   std::optional<std::chrono::milliseconds> timeout = {})
+    -> bool {
+    auto items = std::array<::zmq::pollitem_t, 1>{
+      {{socket.handle(), 0, flags, 0}},
+    };
+    auto infinite = std::chrono::milliseconds(-1);
+    auto ms = timeout ? *timeout : infinite;
+    auto num_events_signaled = ::zmq::poll(items.data(), items.size(), ms);
+    if (num_events_signaled == 0)
+      return false;
+    TENZIR_ASSERT_CHEAP(num_events_signaled > 0);
+    TENZIR_ASSERT_CHEAP((items[0].revents & flags) != 0);
+    return true;
+  }
+
+  engine() = default;
+
+  explicit engine(::zmq::socket_type socket_type)
+    : socket_{ctx_, socket_type}, monitor_{ctx_, socket_} {
+  }
+
+  auto bind(const std::string& endpoint) -> bool {
+    TENZIR_VERBOSE("binding to endpoint {}", endpoint);
+    socket_.bind(endpoint);
+    while (true) {
+      auto event = monitor_.get();
+      TENZIR_DEBUG("got monitor event: {}", render_event(event.event));
+      switch (event.event) {
+        default:
+          break;
+        case ZMQ_EVENT_ACCEPTED:
+          return true;
+        case ZMQ_EVENT_ACCEPT_FAILED:
+        case ZMQ_EVENT_HANDSHAKE_FAILED_AUTH:
+        case ZMQ_EVENT_HANDSHAKE_FAILED_PROTOCOL:
+        case ZMQ_EVENT_HANDSHAKE_FAILED_NO_DETAIL:
+          return false;
+      }
+    }
+    return true;
+  }
+
+  auto connect(const std::string& endpoint) -> bool {
+    TENZIR_VERBOSE("binding to endpoint {}", endpoint);
+    socket_.set(::zmq::sockopt::reconnect_ivl, 250);
+    socket_.connect(endpoint);
+    // The sequence of events we receive when 0mq polls during connecting is as
+    // follows:
+    //     1. ZMQ_EVENT_CONNECT_DELAYED
+    //     2. ZMQ_EVENT_CLOSED
+    //     3. ZMQ_EVENT_CONNECT_RETRIED
+    while (true) {
+      auto event = monitor_.get();
+      TENZIR_DEBUG("got monitor event: {}", render_event(event.event));
+      switch (event.event) {
+        default:
+          break;
+        case ZMQ_EVENT_CONNECTED:
+          return true;
+        case ZMQ_EVENT_HANDSHAKE_FAILED_AUTH:
+        case ZMQ_EVENT_HANDSHAKE_FAILED_PROTOCOL:
+        case ZMQ_EVENT_HANDSHAKE_FAILED_NO_DETAIL:
+          return false;
+      }
+    }
+    return true;
+  }
+
+  ::zmq::context_t ctx_;
+  ::zmq::socket_t socket_;
+  monitor monitor_;
+};
+
+class zmq_loader final : public plugin_loader {
+public:
+  zmq_loader() = default;
+
+  zmq_loader(loader_args args) : args_{std::move(args)} {
+  }
+
+  auto instantiate(operator_control_plane& ctrl) const
+    -> std::optional<generator<chunk_ptr>> override {
+    auto engine = engine::make_source(args_);
+    if (not engine) {
+      TENZIR_ERROR(engine.error());
+      return std::nullopt;
+    }
+    auto make = [&ctrl](class engine engine) mutable -> generator<chunk_ptr> {
+      if (auto message = engine.receive())
+        co_yield *message;
```

Comment:
I "tapped into the raw feed" of my `nic` pipeline, as per the blog post, to receive data, and my `load zmq` pipeline always finishes after receiving the first message. Is this the tapping behavior that you'd expect?

I changed this conditional to a `while` loop, and I got the full-on continuous tap with `nic`. But transmitting non-stream data, e.g. via `from file`, won't terminate the reading pipeline anymore. 

Tired train ride brainstorming: Is there a "terminating message" that the `zmq_saver` can send and the `zmq_loader` only terminates upon receiving that message? Maybe the `engine` can fulfill such behavior?

NOTE: We should probably think about this if a non-Tenzir user wants that kind of feature.

Reply:
> I "tapped into the raw feed" of my nic pipeline, as per the blog post, to receive data, and my load zmq pipeline always finishes after receiving the first message. Is this the tapping behavior that you'd expect?

No, that's not what I'd expect. I'm glad you tested this. I can repro it and will work on a fix.

When I tested this in more depth, I also realized that a disconnected 0mq socket doesn't stop the source from producing after it started once. This needs more accurate connection management.

> I changed this conditional to a `while` loop, and I got the full-on continuous tap with `nic`. But transmitting non-stream data, e.g. via `from file`, won't terminate the reading pipeline anymore.
>
> Tired train ride brainstorming: Is there a "terminating message" that the `zmq_saver` can send and the `zmq_loader` only terminates upon receiving that message? Maybe the `engine` can fulfill such behavior?

This idea goes in the right direction. The proper fix would be processing the 0mq meta events from monitor. There are basically two socket, the monitoring socket and the data socket. If we get an "disconnect" event from the monitoring socket, we should suspend execution.

A `while` loop can't work, but we need the equivalent of a "main loop" that checks both channels. I'll take a stab at it.

Reply:
@Dakostu I was able to fix the issue by fixing the main loop equivalent in saver/loader land. Mind trying it again?

If it works now, just resolve the issue; I've enabled auto-merge.

Reply:
OK, so now I can tap into `nic` data, but receiving pipelines never terminate, which also includes finite data `from file` delivered via `zmq`.

The `zmq_loader`'s `make` loop does not seem to handle the case where this "disconnect" event happens - the `num_peers` does get decreased in the `poll_monitor`, but with no `num_peers`, the generator is just stuck `co_yield`ing empty chunks. We also can't just `co_return`, it would make receiving pipelines non-deterministic; it already runs into that case while establishing connection.

You probably need a second counter `num_disconnected_peers` that increases with each disconnect event - then in the `make` loop, if `num_peers` is 0, check if `num_disconnected_peers` is not 0, then `co_return`, else `co_yield` empty chunks. Good idea?

Reply:
I quickly implemented that idea and it looks like it works now. Should I push it?

Reply:
I would argue that receiving pipelines should never terminate, or at least not by default. The reason is pub/sub semantics: publishers and subscribes come and go. What's stable is the channel in the middle.

I think you're proposing a flag that's similar to `kafka`'s `-e` to exit, right?

Reply:
We discussed this issue in a all and agreed that we are going to leave the issue untouched for this PR, i.e., ship as-is.

The reasoning is that we'd be making strong use case assumptions to terminate a subscriber when the peer count drops to 0, which may work for some but break other use cases. We'd rather wait for user feedback before preemptively adding a `-e` flag.